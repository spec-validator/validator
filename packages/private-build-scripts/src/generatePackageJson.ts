import { Task } from 'just-task'

import { getPackageNamesInBuildOrder } from './buildOrder'
import { read, write } from './readAndWrite'
import getGitVersion from './getGitVersion'

const EXCLUDE = new Set(['devDependencies', 'files'])
const COPY_FROM_PARENT = [
  'license',
  'author',
  'publishConfig',
  'repository',
  'bugs',
  'homepage',
]


// eslint-disable-next-line max-statements
export default (projectPath: string): Task => async () => {
  const parentConfig: Record<string, any> = read('package.json')

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const packageJson: Record<string, any> = read(`${projectPath}/package.json`)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const newPackageJson: Record<string, any> = Object.fromEntries(
    Object.entries(packageJson).filter(it => !EXCLUDE.has(it[0]))
  )

  COPY_FROM_PARENT.forEach(key => {
    if (parentConfig[key]) {
      newPackageJson[key] = parentConfig[key]
    }
  })

  const version: string = getGitVersion()

  newPackageJson.private = packageJson.private || false
  newPackageJson.version = version

  const workspacePackages = new Set(getPackageNamesInBuildOrder())

  if (newPackageJson.dependencies) {
    Object.keys(newPackageJson.dependencies).filter(it => workspacePackages.has(it)).forEach(it => {
      newPackageJson.dependencies[it] = version
    })
  }

  write(`${projectPath}/dist/package.json`, newPackageJson)
}
