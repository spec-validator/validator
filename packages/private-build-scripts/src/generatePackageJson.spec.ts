import generatePackageJson from './generatePackageJson'

jest.mock('./getGitVersion', () => ({
  default: () => '1.1.1',
  __esModule: true,
}))

const write = jest.fn()

jest.mock('./readAndWrite', () => ({
  write: () => write(),
  read: (path: string) => {
    if (path === 'package.json') {
      return {
        'license': 'Apache-2.0',
        'author': {
          'name': 'Anton Berezin',
          'email': 'gurunars@gmail.com',
          'url': 'https://gurunars.com',
        },
        'homepage': 'https://github.com/spec-validator/validator',
        'bugs': {
          'url': 'https://github.com/spec-validator/validator/issues',
        },
        'publishConfig': {
          'registry': 'https://registry.npmjs.org/',
        },
        'repository': {
          'url': 'https://github.com/spec-validator/validator.git',
          'type': 'git',
        },
      }

    } else {
      return {
        'name': 'child-package',
        'version': '0.0.1',
        'dependencies': {
          'pkg1': '0.0.1',
          'pkg2': '0.0.1',
          '3rdpart': '6.6.6',
        },
      }
    }
  },
  __esModule: true,
}))

jest.mock('./buildOrder', () => ({
  getPackageNamesInBuildOrder: () => ['pkg1', 'pkg2'],
  __esModule: true,
}))

test('generatePackageJson', async () => {
  const run = generatePackageJson('parent') as any
  await run()
  expect(write.mock.calls[0][0]).toMatchSnapshot()
})
