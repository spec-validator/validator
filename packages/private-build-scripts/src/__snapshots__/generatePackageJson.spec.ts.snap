// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePackageJson 1`] = `undefined`;
